build-data-capturing-gradle-samples/capture-quality-check-issues/gradle-quality-check-issues.gradle.kts

```
@@ -0,0 +1,129 @@
+                                    val violation = v as NodeChild
+                                    errors.add("${filePath}:${violation.attributes()["lineNumber"]} \u2192 ${violation.attributes()["ruleName"]}")
+                                }
+                            }
+                        }
+                    } else if (ReportingTask.SPOTBUGS.isTask(this)) {
+                        valueName = "Verification SpotBugs"
+                        val bugs = report.getProperty("BugInstance") as NodeChildren
+                        bugs.forEach { b ->
+                            val bug = b as NodeChild
+                            val type = bug.attributes()["type"]
+                            val error = bug.breadthFirst().asSequence().filter { l ->
+                                val line = l as NodeChild
+                                l.name() == "SourceLine"
+                            }.sortedBy { l ->
+                                (l as NodeChild).parent().name()
+                            }.first() as NodeChild
+                            val startLine = error.attributes()["start"]
+                            val endLine = error.attributes()["end"]
+                            val lineNumber = if (endLine == startLine) startLine else "${startLine}-${endLine}"
+                            val className = error.attributes()["classname"]
+                            errors.add("${className}:${lineNumber} \u2192 ${type}")
+                        }
+                    }
+                    errors.forEach { e -> buildScan.value(valueName, e) }
+                }
+            }
+        }
+    }
+}
+
+fun reportFromTask(task: Task): File {
+    if (task is Reporting<*>) {
+        val task = task as Reporting<ReportContainer<*>>
+        return (task.reports.withGroovyBuilder { "xml" { getProperty("destination") } } as Report).getOutputLocation().get().asFile as File
+    } else if (ReportingTask.SPOTBUGS.isTask(task)) {
+        val reports = task.withGroovyBuilder { getProperty("reports") as NamedDomainObjectContainer<*> }
+        val report = (reports.named("XML") as NamedDomainObjectProvider).get() as SingleFileReport
+        return report.getOutputLocation().get().asFile
+    } else {
+        throw IllegalStateException("Unsupported report task: " + task)
+    }
+}
+
+fun appendIfMissing(str: String, suffix: String): String {
+    return if (str.endsWith(suffix)) str else str + suffix
+}
+
+enum class SpotBugsParent {
+    BugInstance, Method, Class
+}
+
+enum class ReportingTask(val className: String) {
+
+    CHECKSTYLE("org.gradle.api.plugins.quality.Checkstyle"),
+    CODENARC("org.gradle.api.plugins.quality.CodeNarc"),
+    SPOTBUGS("com.github.spotbugs.snom.SpotBugsTask");
+
+    fun isTask(task: Task): Boolean {
+        return task::class.java.name.contains(className)
+    }
+
+    companion object {
+        fun isSupported(task: Task): Boolean {
```

i think this would be more idiomatic kotlin if written as an extension property. Does this work?

```suggestion
        val Task.reportingSupported: Boolean {
          return values().any { value -> value.isTask(this) }
        }
```

---

I could not make it work as part of the enum, it works though if defined outside (see updated code).

---

ah, that works and that's what I was thinking!